src/blant.c

```diff
@@ -575,34 +587,74 @@ int RunBlantInThreads(int k, int numSamples, GRAPH *G)
+        }
+        return RunBlantFromGraph(k, numSamples, G, _sampleMethod == SAMPLE_INDEX ? indexSamplingRange : NULL);
+    }
 
-    if (_sampleMethod == SAMPLE_INDEX)
-        Fatal("The sampling method '-s INDEX' does not yet support multithreading (feel free to add it!)");
+    // At this point, _THREADS must be greater than 1.
 
+    //
+    char childrenRunInfinitely = (_outputMode == indexMotifs
+            ||  _outputMode == indexMotifOrbits
+            ||  _outputMode == indexOrbits
+            ||  _outputMode == indexGraphlets
+    )
+    && _sampleMethod != SAMPLE_INDEX
+    && !_window;
 
-    // At this point, _THREADS must be greater than 1.
-    int samplesPerThread = numSamples/_THREADS;  // will handle leftovers later if numSamples is not divisible by _THREADS
 
+    // determine the numSamples per thread. -1 indicates that the child processes will run infinitely until a stop signal from parent is received
+    // Leftovers will be handled later if numSamples is not divisible by _THREAD
+    int samplesPerThread = childrenRunInfinitely ? -1 : numSamples/_THREADS;
+
+    int startNodesPerThread = G->n / _THREADS;
     FILE *fpThreads[_THREADS]; // these will be the pipes reading output of the parallel blants
+
     for(i=0;i<_THREADS;i++) {
         _THREAD_NUM = i;
-        fpThreads[i] = ForkBlant(_k, samplesPerThread, G);
+        if (_sampleMethod == SAMPLE_INDEX) {
+            indexSamplingRange[0] = i * startNodesPerThread;
+            indexSamplingRange[1] = (i + 1) * startNodesPerThread - 1;
+            fpThreads[i] = ForkBlant(_k, numSamples, G, indexSamplingRange);
+        }
+        else {
+            fpThreads[i] = ForkBlant(_k, samplesPerThread, G, NULL);
+        }
     }
 
-
     int threadsDone = 0; // count of how many threads signaled EOF
     int lineNum = 0;
+    int validCount = 0;
+    char perm[maxK+1];
+    TINY_GRAPH *g = TinyGraphAlloc(_k);
+    char line[_numOrbits * BUFSIZ];
+    unsigned Varray[_k];
     do
     {
-	char line[_numOrbits * BUFSIZ];
 	int thread;
 	for(thread=0;thread<_THREADS;thread++)	// read and then echo one line from each of the parallel instances
 	{
 	    if(!fpThreads[thread]) continue; // threads that have finished output have this pointer set to NULL below
-	    char *tmp = fgets(line, sizeof(line), fpThreads[thread]);
-	    assert(tmp>=0);
+        if (childrenRunInfinitely) {
+            if (validCount >= numSamples) { // if there are enough samples, close the fd for current thread
+                fclose(fpThreads[thread]); // after the read end of the pipe is closed, child will receive EPIPE errno when writing to the pipe and then it will terminate
+                fpThreads[thread] = NULL; // signify this pointer is finished.
+                threadsDone++;
+                continue;
+            }
+            else
+                assert(fread(Varray, sizeof(unsigned), _k, fpThreads[thread]) == _k); // read the varray for processing
```

Comment:
Ah. You're sending binary info down the pipe? Have you verified that this reduces CPU usage? It's only worth it if it reduces CPU usage significantly.

Reply:
Previously children processes call ProcessGraphlet to print out formatted graphlets, and I chose to send binary info because I think duplicate checking and formatting in ProcessGraphlet in children processes are unnecessary. I'll try to print out nodes using printf and let the parent tokenize the output and get the node array and process the graphlet, and see if there is any difference in running time or usage.

Reply:
And I'll call NodeSetSeenRecently in children processes so that duplicates will be removed, to avoid the traffic jam in the parent

Reply:
Also, by using write function to write data to the pipe, the children processes are able to determine whether the parent has closed the read end of the pipe by checking the EPIPE errno, and I'm not sure whether I can do similar checking if I use printf